miden-crypto/src/merkle/smt/history/mod.rs

```diff
@@ -0,0 +1,923 @@
+}
+
+impl History {
+    /// Constructs a new history container, containing at most `max_count` historical versions for
+    /// a tree.
+    #[must_use]
+    pub fn empty(max_count: usize) -> Self {
+        let deltas = VecDeque::with_capacity(max_count);
+        Self { max_count, deltas }
+    }
+
+    /// Gets the maximum number of versions that this history can store.
+    #[must_use]
+    pub fn max_versions(&self) -> usize {
+        self.max_count
+    }
+
+    /// Gets the current number of versions in the history.
+    #[must_use]
+    pub fn num_versions(&self) -> usize {
+        self.deltas.len()
+    }
+
+    /// Returns `true` if a new entry can be added without removing the oldest, and `false`
+    /// otherwise.
+    #[must_use]
+    pub fn can_add_version_without_removal(&self) -> bool {
+        self.num_versions() < self.max_versions()
+    }
+
+    /// Returns all the roots that the history knows about.
+    ///
+    /// # Complexity
+    ///
+    /// Calling this method requires a traversal of all the versions and is hence linear in the
+    /// number of history versions.
+    #[must_use]
+    pub fn roots(&self) -> Set<Word> {
+        self.deltas.iter().map(|d| d.root()).collect()
+    }
+
+    /// Returns `true` if `root` is in the history and `false` otherwise.
+    ///
+    /// # Complexity
+    ///
+    /// Calling this method requires a traversal of all the versions and is hence linear in the
+    /// number of history versions.
+    #[must_use]
+    pub fn knows_root(&self, root: Word) -> bool {
+        self.deltas.iter().any(|r| r.root == root)
+    }
+
+    /// Adds a version to the history with the provided `root` and represented by the changes from
+    /// the current tree given in `nodes` and `leaves`.
+    ///
+    /// If adding this version would result in exceeding `self.max_count` historical versions, then
+    /// the oldest of the versions is automatically removed and returned. If no version is pruned,
+    /// the method returns [`None`].
+    ///
+    /// # Gotchas
+    ///
+    /// When constructing the `nodes` and `leaves`, keep in mind that those collections must contain
+    /// entries for the **default value of a node or leaf** at any position where the tree was
+    /// sparse in the state represented by `root`. If this is not done, incorrect values may be
+    /// returned.
+    ///
+    /// This is necessary because the changes are the _reverse_ from what one might expect. Namely,
+    /// the changes in a given version `v` must "_revert_" the changes made in the transition from
+    /// version `v` to version `v + 1`.
+    ///
+    /// # Errors
+    ///
+    /// - [`HistoryError::NonMonotonicVersions`] if the provided version is not greater than the
+    ///   previously added version.
+    pub fn add_version(
+        &mut self,
+        root: Word,
+        version_id: VersionId,
+        nodes: NodeChanges,
+        leaves: LeafChanges,
+    ) -> Result<Option<Delta>> {
```

Comment:
A couple of questions:

- Why do we need to return the oldest version from here? i.e., why not make the return type just `Result<()>`?
- Why not replace `root`, `nodes`, and `leaves` with `MutationSet` as I mentioned in https://github.com/0xMiden/crypto/pull/679#discussion_r2579953876?